### DISCUSSION TITLE
Get text output of a TUI command

### DISCUSSION THREAD

**Discussion by francisco-jimmy** [ansys/pyfluent#1407] [Q&A] [answered]
Hi,

I would like to get, in a python script, the output of a TUI command. For example, the command "define/boundary-conditions/modify-zones/list-zones" prints, in the Fluent GUI, a list of all the model's zones. From pyfluent, the equivalent is solver.tui.define.boundary_conditions.modify_zones.list_zones(). Is it possible to get as a python string the list of zones that is printed in the Fluent GUI when this command is executed?

Thanks in advance.

EDIT:
A workaround is creating a transcript file and reading its contents after executing the TUI command. I would still like to know if there is a more direct way of getting the TUI command text output directly.

**Comment by seanpearsonuk** [comment]
@mkundu1 this seems like it would be nice to have.

**Comment by mkundu1** [comment, marked as answer]
Following is tested on the main branch:
```
import ansys.fluent.core as pyfluent
from contextlib import contextmanager
import io

@contextmanager
def capture_transcript(solver):
    stream = io.StringIO()
    callback_id = solver.transcript.register_callback(stream.write, keep_new_lines=True)
    yield stream
    solver.transcript.unregister_callback(callback_id)

solver = pyfluent.launch_fluent()
solver.file.read_case(file_name="elbow.cas.h5")
with capture_transcript(solver) as stream:
    solver.tui.define.boundary_conditions.modify_zones.list_zones()
transcript = stream.getvalue()
print(transcript)
```

**Reply by etiennelame** [reply]
Hi @mkundu1,
   I am trying you method to get outputas well, but my stream.getvalue() return a blank string. I am using fluent 2022R2, is that related? Beside, my command is solver.execute_tui('define/user-defined/execute-on-demand "cal_rpvar::libudf" \n')

**Reply by etiennelame** [reply]
Hi, 
   I have found the reason.  I run the code **session. transcript.start()** and the function works well. If running **session. transcript.stop()**, the output is not shown in console and the function above return blank list

**Comment by francisco-jimmy** [comment]
Hi,

Thanks a lot! I tested your original solution using the class CaptureTranscript and the solution proposed after the edit and both work well.  Small detail in the original solution: in the __exit__ method we should have self._solver.tui instead of solver.tui right?

**Comment by LeeDuhwan** [comment]
Hi 

This method is not working in 2024R2.

Please help me.

#code -----------------------------------------
```
with capture_transcript(solver) as stream:
    solver.tui.print_license_usage()
transcript = stream.getvalue()
print(transcript)
```

---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[157], line 1
----> 1 with capture_transcript(solver) as stream:
      2     solver.tui.print_license_usage()
      3 transcript = stream.getvalue()

TypeError: 'generator' object does not support the context manager protocol

**Reply by etiennelame** [reply]
hi @LeeDuhwan,
    I have not met your error. If the method went wrong, i usually try restart the transcript service with codes belows:
      solver.transcript.stop() 
      solver.transcript.start("any.trn",True)

**Reply by LeeDuhwan** [reply]
Hi 

Thank you for your comment.

It was resolved.

Thanks.

**Comment by Bjornkullaring** [comment]
Hello! Thanks for @mkundu1 ! It helped me a lot. But it seems that this scripts do now work for Fluent with version 2024. Has anyone tried it in the newset version of Fluent? I am afraid that in v23.1 and v23.2, it worked, however, failed in v24.1. There is no errors but variable transcript is an empty string.

**Reply by LeeDuhwan** [reply]
Hi 

I have tried in V24.2 and this is well working.

What @etiennelame metioned was helpful.

solver.transcript.stop()
solver.transcript.start("any.trn",True)

**Reply by Bjornkullaring** [reply]
Thank you for your reply. I noticed your comment above, and think restarting may help. I will try it in v24.2.

**Reply by Bjornkullaring** [reply]
I checked the method you mentioned. Once faced error, restart the transcrip with the codes below worked for v24.1.
-------------------------------------------
solver.transcript.stop()
solver.transcript.start("any.trn",True)
-------------------------------------------
Thank your very much for your help!!
